File: uia-backend-portal/src/main/java/com/joshuaogwang/uia_backend_portal/auth/AuthenticationService.java
package com.joshuaogwang.uia_backend_portal.auth;

import com.joshuaogwang.uia_backend_portal.config.JwtService;
import com.joshuaogwang.uia_backend_portal.user.Role;
import com.joshuaogwang.uia_backend_portal.user.User;
import com.joshuaogwang.uia_backend_portal.user.UserRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.security.authentication.AuthenticationManager;
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;

@Service
@RequiredArgsConstructor
public class AuthenticationService {
    private final UserRepository userRepository;
    private final PasswordEncoder passwordEncoder;
    private final JwtService jwtService;
    private final AuthenticationManager authenticationManager;

    public AuthenticationResponse register(RegisterRequest request) {
        var user = User.builder()
                .fullName(request.getFullName())
                .phoneNumber(request.getPhoneNumber())
                .email(request.getEmail())
                .password(passwordEncoder.encode(request.getPassword()))
                .role(Role.USER)
                .build();
        userRepository.save(user);
        var jwtToken = jwtService.generateToken(user);
        var userResponse = UserResponse
                .builder()
                .id(user.getId())
                .fullName(user.getFullName())
                .phoneNumber(user.getPhoneNumber())
                .email(user.getEmail())
                .role(user.getRole())
                .build();
        return AuthenticationResponse.builder()
                .token(jwtToken)
                .userResponse(userResponse)
                .build();
    }

    public AuthenticationResponse login(LoginRequest request) {
        authenticationManager.authenticate(
                new UsernamePasswordAuthenticationToken(
                        request.getEmail(), request.getPassword()
                )
        );
        var user = userRepository.findByEmail(request.getEmail()).orElseThrow();
        var jwtToken = jwtService.generateToken(user);
        var userResponse = UserResponse
                .builder()
                .id(user.getId())
                .fullName(user.getFullName())
                .email(user.getEmail())
                .role(user.getRole())
                .build();
        return AuthenticationResponse.builder()
                .token(jwtToken)
                .userResponse(userResponse)
                .build();
    }
}

File: frontend/src/app/pages/authentication/authentication.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { AuthenticationModel } from './authentication.model';
import { Subject, tap } from 'rxjs';
import { Router } from '@angular/router';

interface User {
  id: number;
  fullName: string;
  phoneNumber: string;
  email: string;
  role: string;
}

interface AuthenticationResponse {
  token: string;
  user: User;
}
@Injectable({ providedIn: 'root' })
export class AuthenticationService {

  constructor(private http: HttpClient, private router:Router) {}
  signUp(
    fullName: string,
    email: string,
    phoneNumber: string,
    password: string
  ) {
    return this.http
      .post<AuthenticationResponse>(
        'http://localhost:8080/api/v1/auth/register',
        {
          fullName: fullName,
          email: email,
          phoneNumber: phoneNumber,
          password: password,
        }
      )
  }
  login(email: string, password: string) {
    return this.http.post<AuthenticationResponse>(
      'http://localhost:8080/api/v1/auth/login',
      {
        email: email,
        password: password,
      }
    );
  }
}
